src/meta/src/hummock/compaction/picker/min_overlap_compaction_picker.rs

```diff
@@ -160,194 +166,183 @@ impl NonOverlapSubLevelPicker {
-
-                for other in &overlap_files {
-                    if level_handler.is_pending_compact(&other.sst_id) {
-                        pending_compact = true;
-                        break;
-                    }
-                    overlap_info.update(other);
-                    select_sst_id_set.insert(other.sst_id);
-
-                    all_file_size += other.file_size;
-                    all_file_count += 1;
-                }
-
-                if pending_compact {
-                    break;
-                }
-
-                last_level_index = target_index;
-
-                if overlap_files.is_empty() {
-                    // We allow a layer in the middle without overlap, so we need to continue to
-                    // the next layer to search for overlap
-                    continue;
-                }
-
-                level_select_files[target_index].extend(overlap_files.into_iter());
-                select_level_count += 1;
+                overlap_info.update(other);
+                select_sst_id_set.insert(other.sst_id);
+                all_file_size += other.file_size;
             }
 
             if pending_compact {
                 break;
             }
 
-            // check reverse overlap
-            for reverse_index in (0..last_level_index).rev() {
-                let target_tables = &l0[reverse_index].table_infos;
+            let mut extra_overlap_levels = vec![];
+
+            for reverse_index in (0..target_index).rev() {
+                let target_tables = &levels[reverse_index].table_infos;
+                // It has select all files in this sub-level, so it can not overlap with more files.
+                if ret.sstable_infos[reverse_index].len() == target_tables.len() {
+                    continue;
+                }
                 let overlap_files = overlap_info.check_multiple_overlap(target_tables);
                 let mut extra_overlap_sst = Vec::with_capacity(overlap_files.len());
                 for other in overlap_files {
                     if level_handler.is_pending_compact(&other.sst_id) {
                         pending_compact = true;
                         break;
                     }
-
                     if select_sst_id_set.contains(&other.sst_id) {
                         // Since some of the files have already been selected when selecting
                         // upwards, we filter here to avoid adding sst repeatedly
                         continue;
                     }
 
                     all_file_size += other.file_size;
-                    all_file_count += 1;
-
                     overlap_info.update(&other);
                     select_sst_id_set.insert(other.sst_id);
                     extra_overlap_sst.push(other);
                 }
 
-                if pending_compact {
+                // If the current
+                if extra_overlap_sst.is_empty() {
```

Comment:
we could not break when someone level extra_overlap_sst empty, It is possible that the column does not contain data at that level，but data exists in the subsequent level.

Reply:
fixed